tern/extensions/scancode/executor.py

```diff
@@ -114,13 +114,17 @@ def add_file_data(layer_obj, collected_files):
 
 class Scancode(Executor):
     '''Execute scancode'''
-    def execute(self, image_obj):
+    def execute(self, image_obj, redo=False):
         '''Execution should be:
             scancode -ilpcu --quiet --json - /path/to/directory
         '''
         for layer in image_obj.layers:
             # load the layers from cache
-            common.load_from_cache(layer)
+            common.load_from_cache(layer, redo)
+            # reading files again in case of redo
+            if redo:
+                logger.debug('Reading files in filesystem...')
+                layer.add_files()
```

Comment:
This bit of logic is always true if you call `common.load_from_cache(layer)` even with `redo is True`.

Reply:
I mean this doesn't have to change. `load_files_from_cache` has this logic and gets called in `load_from_cache` outside the `redo` check. The reason for this is that tern reads the files when loading the images by default, so the files always exist. You're repeating the same operation here.

Reply:
I guess it is not getting called in `load_files_from_cache`. If I pass `redo=True`, the method will just return `loaded=False`.

`def load_from_cache(layer, redo=False):
    '''Given a layer object, check against cache to see if that layer id exists
    if yes then load any relevant layer level information. The default
    operation is to not redo the cache. Add notices to the layer's origins
    matching the origin_str'''
    loaded = False
    if not redo:
        # check if packages are available in the cache
        if load_packages_from_cache(layer):
            loaded = True
        # load some extra properties into the layer if available
        if layer.fs_hash in cache.get_layers():
            layer.files_analyzed = cache.cache[layer.fs_hash]['files_analyzed']
            layer.os_guess = cache.cache[layer.fs_hash]['os_guess']
            layer.pkg_format = cache.cache[layer.fs_hash]['pkg_format']
        load_files_from_cache(layer)
    return loaded
`
That is why I though to add.

Reply:
Yes, it's meant to work that way - load the files from the cache regardless of whether redo is true or false.
```
def load_from_cache(layer, redo=False):                                         
 41     '''Given a layer object, check against cache to see if that layer id exists 
 42     if yes then load any relevant layer level information. The default          
 43     operation is to not redo the cache. Add notices to the layer's origins      
 44     matching the origin_str'''                                                  
 45     loaded = False                                                              
 46     if not redo:                                                                
 47         # check if packages are available in the cache                          
 48         if load_packages_from_cache(layer):                                     
 49             loaded = True                                                       
 50         # load some extra properties into the layer if available                
.....                                  
 55         load_files_from_cache(layer)  <-- we do not set loaded to True because the files always need to be loaded from cache.                                      
 56     return loaded   
```

Reply:
Actually, now that you made this change, this line can be moved out of the `if not redo:` check.

Reply:
When I ran with `redo = True` condition in `executor` then it was not able to read anything at all and the report was coming empty I guess because `load_files_from_cache` was not get called by default So I added this condition there. Maybe I can remove the condition and make `load_files_from_cache` to get called by default.